chore: rabbitmq 설정

eTaphee opened the pull request:
<!-- 필요한 경우 팀원들과 의논하여 양식을 변경할 수 있습니다. -->
<!-- PR은 상세히 적어주시는게 좋습니다. -->
<!-- 이해를 돕기위한 이미지를 써도 좋습니다. -->

## Changes
<!-- 어떤점이 변경되었는지 적어주세요. -->
- Rabbit MQ를 설치하고 실행할 수 있도록 docker 명령을 실행하는 쉘스크립트 파일을 작성했습니다.
  - 도커이미지는 관리 UI가 포함된 management-alpine 태그를 사용합니다.
  - 통신은 5672포트를 사용하며, 15672 포트는 웹 관리 UI를 위한 접속 포트입니다.
  - 기본 접속 정보는 guest/guest 입니다.
```sh
docker pull rabbitmq:management-alpine
docker run -d --name rabbitmq-banzzokee -p 5672:5672 -p 15672:15672 --restart=unless-stopped rabbitmq:management-alpine
```
- MQ에 전송되는 메시지는 `BaseMessage`를 상속받아 구현합니다.
```java
// 분양 게시글 등록 이벤트
public class AdoptionRegisteredEvent extends BaseMessage<AdoptionDto> {
  public static AdoptionRegisteredEvent from(AdoptionDto adoption) {
    return AdoptionRegisteredEvent.builder()
        .payload(adoption)
        .build();
  }
}

// AdoptionService.cs
private final ApplicationEventPublisher eventPublisher;

public void registerAdoption(AdoptionRegisterRequest request, Long userId) {
  Adoption adoption = adoptionRepository.save(request.toEntity());
  eventPublisher.publish(AdoptionRegisteredEvent.from(AdoptionDto.fromEntity(adoption));
}

@Service
@RequiredArgsConstructor
public class AdoptionRegisteredEventHandler {

  private final RabbitMqMessageProducer rabbitMqMessageProducer;

  @Async
  @TransactionalEventListener(classes = AdoptionRegisteredEvent.class, phase = AFTER_COMMIT)
  public void handle(AdoptionRegisteredEvent event) {
    rabbitMqMessageProducer.publish("adoption.created", event);
  }
}
```

**설정 정보**
![rabbitmq drawio](https://github.com/banzzokee/banzzokee-back/assets/43163409/c89bdb51-aced-42cf-b1b6-7dee478c30a1)
### Producer
- API 서버가 Producer의 역할을 수행합니다.
- Producer가 생성하는 메시지 종류는 다음과 같습니다.
  - ES 동기화를 위한 데이터(분양글, 리뷰글 CUD)
  - FCM을 통해 보낼 알림 데이터
  - FCM 구독 설정을 위한 데이터
### Rabbit MQ
- Producer로 부터 publish 된 메시지는 exchange에 먼저 도달합니다.
- exchange는 publish에서 지정한 routing key를 확인하고 룰에 따라 각 큐에 전송합니다.
  - 바인딩 룰은 `RabbitMqConfig.java` 파일을 확인하면 됩니다.
- 용도 별 queue를 관리합니다.
  - queue.sync.elasticsearch.adoption: 입양글 동기화
  - queue.sync.elasticsearch.review: 리뷰글 동기화
  - queue.notify.fcm.adoption: 입양글 작성 알림
  - queue.notify.fcm.review: 리뷰글 작성 알림
  - queue.notify.fcm.chat: 채팅 알림
  - queue.manage.fcm.subscription: fcm 토픽 구독 관리(팔로우, 북마킹)
  - queue.manage.fcm.token: fcm 토큰 등록 시 토픽 구독 등록(입양글,리뷰글)
### Consumer
- 각 컨슈머가 큐를 구독하여 이벤트를 처리합니다.

## Background
<!-- 이 PR이 진행된 배경을 적어주세요. -->
다음과 같은 목적으로 외부 메시지 브로커를 사용할 필요가 있습니다.
- FCM 메시지 처리
  - 게시글 알림
  - 리뷰 알림
  - 채팅 알림
- 분양글, 리뷰글 CUD 작업 ElasticSearch 동기화

위 기능들을 하나의 어플리케이션에서 수행한다면 수정 발생 시 모두 내렸다 올려야하는 문제가 있습니다.
향후 멀티모듈 구성으로 프로세스를 분리할 필요가 있습니다.

## Discuss
<!-- 토론할 내용이 있다면 적어주세요. -->
- STOMP에 대한 queue 설정
- produce 실패 시 재시도 전략
- consume 실패 시 재시도 전략 (dead queue)

## Issues
<!-- 관련된 이슈를 #{issueNumber}를 통해 태그해주세요. -->
<!-- ex) 해결한 이슈: #1, #2 -->
<!-- ex) 구현이 필요한 이슈: #3, #4 -->
#77 

## Execute
<!-- 실행된 결과에 대해 적어주세요. -->
<!-- 어떻게 테스트를 진행했는지, 어떤 검토를 거쳤는지 적어주시면 좋습니다. -->
<!-- 테스트 코드는 중요합니다! -->
POSTMAN으로 토큰 등록 API를 호출하고,
RabbitMQ 관리 UI를 통해 라우팅되는 것과 메시지를 확인했습니다.
<img width="1187" alt="image" src="https://github.com/banzzokee/banzzokee-back/assets/43163409/88a4aa2d-616d-41d7-b4c8-ba759581c8a6">

테스트할 수 있는 API 및 예시 코드는 작성하지 않았습니다.

## Reference
<!-- 참조한 레퍼런스가 있다면 적어주세요. -->

eTaphee committed:
chore: rabbitmq 설정

eTaphee committed:
chore: build.gradle

eTaphee committed:
Merge branch 'develop' into feature/rabbitmq-setting

jnjeaaaat commented:
수고 많으셨습니다~!

eTaphee committed:
feature: fcm 토큰 produce, consume 구현

FCM 토큰 등록 시, 보호소 목록 및 북마크 목록에 대해 토픽 구독하고 구독 정보를 저장한다.

eTaphee committed:
test: fcm 토큰 등록 이벤트 처리 테스트 작성

eTaphee committed:
feature: unsubscribeFromTopic 메서드 추가

nadoran78 reviewed (approved):
메시지 큐 기능 자체를 처음 봤는데 많은 참고 되었습니다. 고생하셨습니다~

eTaphee committed:
chore: 토픽 관리 큐 이름 변경, 바인딩 룰 추가

- queue.manage.fcm.subscription -> queue.manage.fcm.topic 변경
- 바인딩 룰이 누락되어 추가

eTaphee committed:
refactor: BaseMessage 리팩터링

- 라우팅 키를 이벤트 핸들러에서 명시해서 보내는 방법에서 이벤트 클래스에서 getRoutingKey() 를 오버라이딩 하는것으로 변경
- 메시지 포맷이 같으므로, 각각의 메시지를 처리하는 핸들러 대신 BaseMessage를 처리하는 MessageHandler 추가
- MessageHandler 추가로 FcmTokenRegisteredEventHandler 삭제

eTaphee committed:
feature: 보호소 사용자 팔로우, 언팔로우  produce, consume 구현

- 팔로우, 언팔로우 행위는 결국 topic를 subscribe/unsubscribe하는 행위이므로 FcmTopicStatusChangeEvent 클래스로 추상화
- FcmTopicStatusChangeEvent 클래스를 상속받아 상세 이벤트 클래스 구현
- FcmTopicStatusChangeEvent 이벤트 처리 핸들러 구현
- FcmTokenRegisteredEventListener 클래스도 topic를 subscribe/unsubscribe하기 떄문에 같은 Consumer 클래스로 묶음

eTaphee committed:
refactor: 패키지 명 수정 누락 반영

eTaphee committed:
test: 사용자 팔로우 테스트 수정 및 언팔로우 테스트 케이스 작성

- 팔로우 시 발행되는 ShelterUserFollowedEvent 이벤트 검증
- 언팔로우 이벤트 추가
  - 팔로우 정보가 없으면 delete, publishEvent 호출하지 않음
  - 팔로우 정보가 있으면 delete, publishEvent 호출

eTaphee committed:
test: 토픽 상태 변경 처리 테스트 작성

- 액션이 SUBSCRIBE 시 subscribeTopicsAndSave 검증
- 액션이 UNSUBSCRIBE 시 unsubscribeTopicsAndDelete 검증

eTaphee committed:
fix: 자잘한 버그 및 기타 수정

- follower, followee 헷갈려서 바꿔쓴 코드 수정
- topic 필드가 있고 getTopic()이 이름이 겹처서 역직렬화 안되는 문제가 있어서 topicCategory로 변경

eTaphee committed:
refactor: event 클래스 단순화

불필요한 상속 클래스 제거

eTaphee committed:
Merge pull request #87 from banzzokee/feature/produce-and-consume-shelter-followed-event

feature: produce and consume shelter followed event

eTaphee committed:
Merge pull request #83 from banzzokee/feature/produce-and-consume-fcm-token-registered-event

feature: produce and consume fcm token registered event

eTaphee committed:
Merge branch 'develop' into feature/rabbitmq-setting

eTaphee committed:
test: DataJpaTest 테스트 QueryDSLConfig 추가

ChatRoomServiceTest 테스트 클래스에서 @DataJpaTest를 사용하지만, QueryDslConfig가 로드되지 않아 문제가 발생
